### ISSUE TITLE
[Question] How may one direct pip-compile + pip-sync to use `conda` instead of `pip` for some packages? 

### ISSUE THREAD

**Issue by davidleejy** [jazzband/pip-tools#1833]
Suppose one would like to perform the equivalent of the following in setting up a python environment?

```bash
$ conda install ipykernel
$ conda install -c conda-forge ipywidgets
$ pip install numpy
```

How would one go about doing this using `pip-tools`?

**Comment by AndydeCleyre** [comment]
Curious: what's the advantage in this case over:

`requirements.in`:

```ini
ipykernel
ipywidgets
numpy
```

```console
$ pip-compile
$ pip-sync
```

?

**Comment by davidleejy** [comment]
Thanks for the suggestion. I was using ipykernel and ipywidgets as examples 😀

While pip installing ipykernel and ipywidgets might work, it has been my experience that some packages on the conda repo, though identically named as their pypi counterparts, have differences in source code. If my memory serves me well, I think I've experienced this with matplotlib (threw an exception that wasn't in the official codebase). I have, in the past, resolved problems encountered during the installation of, or when using, a pypi package by replacing it with its conda counterpart.

Would be neat if pip-tools happens to handle specifications about which server a package should be installed from.

I suppose that this feature could also be useful for installing packages from private package hosting servers or a server that's on an enterprise's intranet 🙂

**Comment by groodt** [comment]
conda and pip are different ecosystems with different metadata. They can’t be mixed safely into a single environment and that’s acknowledged by both ecosystems. 

conda is a system environment manager (ie. not only python packages).

pip is only a python package resolver and installer

They have different metadata and purposes, so I don’t see how this would fall under the scope of pip-tools. 

conda already has “conda lock” for creating an environment specification.

**Comment by davidleejy** [comment]
@groodt Thanks for the detailed comment. Yes, I can see where you're coming from. Would I be interpreting your first paragraph, "conda and pip are different ecosystems [...] and that's acknowledged by both ecosystems." correctly if I took that to mean that the pip-tools resolver wouldn't be able to give an accurate take on the compatibility of packages installed from different channels (channels meaning conda, pypi, etc.), due to each channel's definition of compatibility being limited to its own channel? So, pypi packages each lists dependencies that are contingent on the these dependencies being installed from pypi?

Thank you also for suggesting conda lock. At the moment, it appears to possess some functionality similar to pip-tools, and seems promising for users and teams with projects that rely on conda.

While I agree that it is generally preferable to avoid installing a mix of packages from conda and pypi into a single environment, it has been my experience that sometimes it appears all right to mix _certain_ conda packages with pip ones; e.g., jupyter related packages from conda (development environment related) + other packages from pypi. I'm guessing that people with such a set-up might have to combine two tools -- one that has a conda resolver (like conda-lock I'm guessing) and pip-tools -- with a script that triggers commands in both to do the necessary.

I feel more or less ready to close this issue if the maintainers so prefer.

**Comment by groodt** [comment]
Conda is solving a broader problem and a conda lockfile file will be able to capture more of the environment than pip or pip-tools will.

If you are already committed to conda and a happy conda user, then the approach with python packages is to use conda-forge to get the packages into conda https://conda-forge.org/docs/

Most common packages that are actively are already there. That demonstrates one big difference with conda, they rebuild packages from source to make them compatible with the rest of the conda ecosystem.

You may have some luck with pure python packages that work from pip with conda (or vice-Verda), but it would likely be quite fragile and probably only works with pure python packages. 

Mixing pip and conda would be like mixing npm and Homebrew ecosystems. The scope is very different. I do of course acknowledge that conda does have a strong niche with python developers.

**Comment by sirosen** [comment]
As noted above, this is very firmly out of scope. I don't see any benefit in keeping this open, so I'm closing.
